### ISSUE TITLE
Fix network interface enumeration on Mono

### ISSUE THREAD

**Issue by NattyNarwhal** [MonoOni/binarydist#28]
Right now Mono can't enumerate network interfaces; it assumes none are installed. What's happening is Mono assumes in the default Unix interface enumerator class that it can list them off with `getifaddrs`, which isn't present under AIX, and is available in a different form with a different name under i. What can be done is writing a backend or two using `SIOCGIFCONF` based enumeration for AIX; and perhaps one using `Qp2getifaddrs` under i. (Alternatively, we could point to a wrapper library, but this is a pain.)

This is of relatively high importance because the Socket class uses these to enumerate supported network protocols; and that is used by things like DNS. Fixing this will enable a huge swath of tests under AIX CI. Right now, upstream is in the middle of refactoring this, so I don't know if I should wait for them to finish or not.

In the mean time, our binary builds supply a patch to nop out the check for interfaces in Sockets init.

**Comment by kadler** [comment]
FYI, we have some code which implements a `getifaddrs` shim for PASE already that we're looking at open sourcing. I'm not sure if the same would work on AIX, though.

**Comment by NattyNarwhal** [comment]
Is that the same as `Qp2getifaddrs`? I know rsyslog implements a shim.

I have a `SIOCGIFCONF` based implementation of `NetworkInterface` running on AIX, but I haven't tested it under PASE yet. I assume it'd work except for maybe MTUs, but the incongruency between the ioctl and `getifaddrs`/`if_nameindex` functions for device names is annoying and prevents me from getting the index; not essential for the class I implemented but other stuff like in CoreFX PAL really wants the index. It seems the ioctl on AIX returns `AF_LINK` containing `sockaddr_dl`s, but not under PASE, so that's not a workaround. Is there another method perhaps?

Interestingly, all methods seem to return the same stuff regardless of PASE or ILE.

**Comment by kadler** [comment]
No, it's through a different method and uses neither `Qp2getifaddrs` nor `SIOCGIFCONF`

**Comment by NattyNarwhal** [comment]
An [attempt](https://github.com/mono/mono/pull/9454) but upstream is hesitant due to a future refactor Real Soon Now™.

**Comment by NattyNarwhal** [comment]
Upstream has merged the AIX patch for now. It won't work 100% on i, but enough to omit a patch. When upstream works on importing CoreFX NetworkInterface, I have an idea of the interface to use instead that works properly on both AIX and i.
